spinel/pcap.py

```diff
@@ -46,62 +47,64 @@
 LQI_LEN = 1
 
 
-
-def crc( s ):
+def crc(s):
     # Some chips do not transmit the CRC, here we recalculate the CRC.
 
     crc = 0
     # remove the last 2 bytes
     for c in s[:-2]:
-        c = ord(c)
-        q = (crc ^ c) & 0x0f
+        q = (crc ^ c) & 0x0F
         crc = (crc >> 4) ^ (q * 0x1081)
-        q = (crc ^ (c >> 4)) & 0x0f
+        q = (crc ^ (c >> 4)) & 0x0F
         crc = (crc >> 4) ^ (q * 0x1081)
-    msb = chr( 0x0ff & (crc >> 8) )
-    lsb = chr( 0x0ff & (crc >> 0) )
-    s = s[:-2] + lsb + msb
+
+    # lsb
+    s[-2] = 0xFF & (crc >> 0)
+    # msb
+    s[-1] = 0xFF & (crc >> 8)
     return s
 
+
 class PcapCodec(object):
     """ Utility class for .pcap formatters. """
 
     @classmethod
     def encode_header(cls, dlt):
         """ Returns a pcap file header. """
         cls._dlt = dlt
-        return struct.pack("<LHHLLLL",
-                           PCAP_MAGIC_NUMBER,
-                           PCAP_VERSION_MAJOR,
-                           PCAP_VERSION_MINOR,
-                           0, 0, 256,
-                           cls._dlt)
+        return struct.pack("<LHHLLLL", PCAP_MAGIC_NUMBER, PCAP_VERSION_MAJOR, PCAP_VERSION_MINOR, 0, 0, 256, cls._dlt)
 
     @classmethod
     def encode_frame(cls, frame, sec, usec, options_rssi, options_crc, metadata=None):
         """ Returns a pcap encapsulation of the given frame. """
         # write frame pcap header
         TLVs_length = TLVS_LENGTH_DEFAULT
-        
+
+        frame = bytearray(frame)
+
         if options_crc:
             frame = crc(frame)
             TLVs_length += 8
-            
+
         if options_rssi:
-            if (cls._dlt == DLT_IEEE802_15_4_TAP):
+            if cls._dlt == DLT_IEEE802_15_4_TAP:
                 TLVs_length += 16
             else:
-                # TI style FCS format: replace the last two bytes with RSSI and LQI
-                frame = frame[:-2] + chr(metadata[0] & 0xff) + chr(metadata[3][1] & 0xff)
-            
-        if (cls._dlt == DLT_IEEE802_15_4_TAP):
+                # TI style FCS format: replace the last two bytes (should be FCS) with RSSI and LQI and always
+                # assume FCS right
```

Comment:
should keep wrong FCS if FCS is wrong?

Reply:
I may not understand what you mean.
However, in Wireshark 3.0 and later versions, we can use TAP to display CRC, RSSI and LQI at the same time. However, in Wireshark 2.x version, CRC and RSSI occupy the same position and cannot be displayed at the same time.